### PR TITLE
Improve height matching

### PR DESCRIPTION
**Pull request by keenon**
This PR improves the quality of results in several ways:

- Primarily, we add a `fitter.setExplicitHeightPrior(heightM, weightOfPrior = 1e3)` method to `MarkerFitter` that can be used to add a term to track the subject height explicitly (at a much higher weight than the rest of the anthropometric prior)
  - This term gets incorporated both in the initial scale+fit steps, as well as the final bilevel optimization
- We also improve the quality of the joint center and joint axis fits
- Also minor improvements to the initial IK steps, with better root positions during random restarts

### CODE REVIEW

File: dart/biomechanics/MarkerFitter.cpp [hunk 1]
```diff
@@ -710,16 +714,37 @@ MarkerFitter::MarkerFitter(
...
+    loss += anthroRegularization;
+
+    // 4. Regularize the height of the body, if we've got a height prior
+    s_t heightRegularization = 0.0;
+    if (mHeightPriorWeight > 0.0)
     {
-      state->bodyScalesGrad.setZero();
+      Eigen::VectorXs pose
+          = Eigen::VectorXs::Zero(this->mSkeleton->getNumDofs());
+      pose(4) = 0.94;
```
Comment (nickbianco) [inline dart/biomechanics/MarkerFitter.cpp:730]: This seems to be hard-coding the default y-position of the model? Will the loss term not function correctly if a generic model with a different y-pos convention is used?

File: dart/biomechanics/MarkerFitter.cpp [hunk 2]
```diff
@@ -710,16 +714,37 @@ MarkerFitter::MarkerFitter(
...
+      s_t height = this->mSkeleton->getHeight(pose);
+      Eigen::VectorXs heightGrad
+          = this->mSkeleton->getGradientOfHeightWrtBodyScales(pose);
+
+      s_t diff = height - mHeightPrior;
+      heightRegularization = mHeightPriorWeight * diff * diff;
+
+      // 3.3. Translate gradients from vector back to matrix form for the state
```
Comment (nickbianco) [inline dart/biomechanics/MarkerFitter.cpp:738]: Minor, but the numbering on this comment should be updated.

File: dart/biomechanics/MarkerFitter.cpp [hunk 3]
```diff
@@ -4560,6 +4598,13 @@ ScaleAndFitResult MarkerFitter::scaleAndFit(
       }
     }
 
+    // Get the neutral pose of the skeleton where we'll measure the height
+    Eigen::VectorXs skeletonHeightMeasurementPose
+        = Eigen::VectorXs::Zero(skeleton->getNumDofs());
+    skeletonHeightMeasurementPose(5) = 0.94;
```
Comment (nickbianco) [inline dart/biomechanics/MarkerFitter.cpp:4604]: Why does this use index `5`, whereas above you use index `4`?

File: dart/biomechanics/MarkerFitter.cpp [hunk 4]
```diff
@@ -5758,7 +5859,8 @@ std::shared_ptr<SphereFitJointCenterProblem> MarkerFitter::findJointCenter(
   for (int i = 0; i < mJointSphereFitSGDIterations; i++)
   {
     Eigen::VectorXs grad = problem->getGradient();
-    Eigen::VectorXs newAccum = accum + grad.cwiseProduct(grad);
+    Eigen::VectorXs newAccum
+        = decay * accum + (1.0 - decay) * grad.cwiseProduct(grad);
```
Comment (nickbianco) [inline dart/biomechanics/MarkerFitter.cpp:5863]: Is `decay` supposed to be iteratively updated? If not, this sum is now much more biased towards the current value of `accum`. Maybe that's intentional, just trying to understand.

File: unittests/unit/test_MarkerFitter.cpp [hunk 5]
```diff
@@ -1668,6 +1670,12 @@ std::vector<MarkerInitialization> runEngine(
       0,
       2);
 
+  Eigen::VectorXs pose = Eigen::VectorXs::Zero(standard.skeleton->getNumDofs());
+  pose(4) = 0.94;
```
Comment (nickbianco) [inline unittests/unit/test_MarkerFitter.cpp:1674]: See comments about indexing above.

**Review by nickbianco — CHANGES_REQUESTED** [review]
There's a couple of minor things to double check, but should be easy to update.

File: dart/biomechanics/MarkerFitter.cpp (hunk 1 above)
Comment (keenon) [reply dart/biomechanics/MarkerFitter.cpp:730]: This is an excellent catch. I was just blindly copying the way the `ANSUR_metrics.xml` defines the height metric, but that's dumb. It shouldn't matter what the root translation of the model is when measuring height, so I've deleted these offset lines.

File: unittests/unit/test_MarkerFitter.cpp (hunk 5 above)
Comment (nickbianco) [reply unittests/unit/test_MarkerFitter.cpp:1674]: If you care to you can remove this, but as you said, it shouldn't matter

**Review by nickbianco — APPROVED** [review]
LTGM. One manual pelvis height pose was left in the new test, but it's up to you if you want to remove it before merging.
